geniza/corpus/management/commands/merge_joins.py

```diff
@@ -18,12 +20,28 @@
 from geniza.corpus.models import Document
 
 
+@dataclass
+class MergeGroup:
+    """Group of Documents that are ready to be merged: candidates in `to_merge`
+    will be combined with `primary` using Document.merge_with, creating log
+    entries to document the merge reasoning using `rationale`."""
+
+    primary: int = None  # pgpid
+    to_merge: List[int] = field(default_factory=list)  # list of pgpid
+    rationale: str = ""
```

Comment:
the reason for the seemingly useless defaults here is so that later on we can do a `defaultdict(MergeGroup)` and assign to the inner `to_merge` list without needing to initialize each one manually. that way, the rows from the .csv can be read in any order (for example, a merge candidate could be listed before the group's primary) and we will still get the right data structure

Reply:
I looked at this briefly at the end of the day yesterday and it wasn't making much sense — maybe because I'm not familiar with dataclass (seems interesting, though).

This seems more complicated than we nee. Why not just load the spreadsheet rows into a list of dicts and then use django regroup? Then you don't have to worry about getting them in order, you will always process them as complete group.

Reply:
should've mentioned this at standup, sorry – what I wanted to do was process the rows (in whatever order they occur in the spreadsheet) in one pass and then be done; no sorting or grouping needed afterwards. `defaultdict` seems like a natural choice because we already have the keys specified, and i potentially need to assign/append to the `to_merge` property without it being initialized as a list first. the value type for the `defaultdict` needs to be mutable, so I picked a solution that would give me:
- `__init__` for free, so that `to_merge` would be an empty list
- `__repr__` for free, because i wanted to see a representation in the console that was compact like a tuple
- access to members as attributes, like `MergeGroup.primary`
- type annotation for the members
- a place to document what the record is supposed to do
- associated behavior (the `MergeGroup.merge()` function)

i don't think this adds too much complexity in return for being explicit about the data type, and i think it'll help in tests.